### PR TITLE
Short document that recap the difference python/IPython

### PR DESCRIPTION
**Pull request by Carreau**
@fperez  want a short document that recap all the "languages" difference IPython/Python. 

Maybe @willinc can rephrase that a bit better than I ? 

@fperez thoughts ?

### CODE REVIEW

**Comment by minrk** [comment]
@Carreau thanks for this! It's a great summary. I'll let @fperez review, since he has the most experience with defining IPython's syntax extensions.

File: docs/source/interactive/python-ipython-diff.rst [hunk 1]
```diff
@@ -0,0 +1,251 @@
+==================
+IPython vs IPython
```
Comment (damianavila) [inline docs/source/interactive/python-ipython-diff.rst]: IPython against itself? :wink:

File: docs/source/interactive/python-ipython-diff.rst [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+==================
+
+This document is meant to highlight the main differences between the python
```
Comment (damianavila) [inline docs/source/interactive/python-ipython-diff.rst]: Python, in uppercase here, I think...

File: docs/source/interactive/python-ipython-diff.rst [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        [('a', 5), ('b', 4), ('c', 3)]
+        >>> sorted(c)                       # list all unique elements
+        ['a', 'b', 'c', 'd', 'e']
+        >>> ''.join(sorted(c.elements()))   # list elements with repetitions
+        'aaaaabbbbcccdde'
+        ...
+
+
+
+If you are looking for an object, the use of wildcards ``*`` in conjonction
```
Comment (damianavila) [inline docs/source/interactive/python-ipython-diff.rst]: conjunction

File: docs/source/interactive/python-ipython-diff.rst [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+with question mark will allow you to search current namespace for object with
+mathcing names:
```
Comment (damianavila) [inline docs/source/interactive/python-ipython-diff.rst]: matching

File: docs/source/interactive/python-ipython-diff.rst [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    int
+    print
+
+
+Shell Assignment
+================
+
+
+When doing interactive computing it is common to need to access the underlying shell.
+This is doable through th use of the exclamation mark ``!`` (or bang).
```
Comment (damianavila) [inline docs/source/interactive/python-ipython-diff.rst]: the

File: docs/source/interactive/python-ipython-diff.rst [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    In[1]: !cd /var/etc
+
+Or edit file:
+
+.. code-block::
+
+    In[1]: !mvim myfile.txt
+
+
+The line after the bang can can call any program installed in the underlying
```
Comment (damianavila) [inline docs/source/interactive/python-ipython-diff.rst]: double `can`

File: docs/source/interactive/python-ipython-diff.rst [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+    In[2]: !mv $file {file.upper()}
+
+
+The bang can also be present in the right hand side of an assignment, just
+after the equal sign, or separated from it by a white space. In which case the
+standard output of the command after the bang ``!`` will be spilt out into line
+in a list-like object (:see:`IPython Slist`) and assign to the left hand side. 
+
+This allow you for example to put th list of files of the current working directory in a variable:
```
Comment (damianavila) [inline docs/source/interactive/python-ipython-diff.rst]: the

File: docs/source/interactive/python-ipython-diff.rst [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        raw = !echo $backup $file
+        !cp $file {file.split('.')[0]+'.bak'}
+
+
+Magics
+------
+
+Magics function are often present in the form of shell-like syntax, but are
+under the hood python function. The syntax and assignment possibility are
+similar to the one with the bang (``~``) syntax, but with more flexibility and
```
Comment (damianavila) [inline docs/source/interactive/python-ipython-diff.rst]: ``!``

**Comment by damianavila** [comment]
Just commented on some typos... probably there is more improvements to do in the phrasing but I will let that to other with English-native knowledge... Thanks @Carreau for working on this!

File: docs/source/interactive/python-ipython-diff.rst [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+    ?object
+
+
+.. code-block::
+
+    object?
+
+.. code-block::
+
+    *patern*?
```
Comment (njsmith) [inline docs/source/interactive/python-ipython-diff.rst]: Missing `t`

File: docs/source/interactive/python-ipython-diff.rst [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    Type:      type
+
+
+A double question mark will try to pull out more information about the object,
+and if possible display the python source code of this object. 
+
+.. code-block::
+
+    In[1]: import collections
+    In[2]: collection.Counter?
```
Comment (njsmith) [inline docs/source/interactive/python-ipython-diff.rst]: Should be double `??`, no?

File: docs/source/interactive/python-ipython-diff.rst [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+Or edit file:
+
+.. code-block::
+
+    In[1]: !mvim myfile.txt
+
+
+The line after the bang can can call any program installed in the underlying
+shell, and support variable expansion in the form of ``$variable`` or ``{variable}``.
+The later for of expansion support arbitrary python expression:
```
Comment (njsmith) [inline docs/source/interactive/python-ipython-diff.rst]: "form", "supports"

File: docs/source/interactive/python-ipython-diff.rst [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+.. code-block::
+
+    In[1]: file = 'myfile.txt'
+
+    In[2]: !mv $file {file.upper()}
+
+
+The bang can also be present in the right hand side of an assignment, just
+after the equal sign, or separated from it by a white space. In which case the
+standard output of the command after the bang ``!`` will be spilt out into line
```
Comment (njsmith) [inline docs/source/interactive/python-ipython-diff.rst]: split, lines

**Comment by minrk** [comment]
I just discovered a new syntax extension we have that I'm not sure I ever knew about: `/` triggers autocall.

```
In [1]: def foo(): print("FOO")

In [2]: foo
Out[2]: <function __main__.foo>

In [3]: /foo
FOO
```

**Comment by fperez** [comment]
Yup, fresh new feature, introduced I think around 2001 ;)

**Comment by fperez** [comment]
Note, [this notebook, "beyond plain Python"](https://github.com/ipython/ipython/blob/master/examples/IPython%20Kernel/Beyond%20Plain%20Python.ipynb), is the one that best summarizes our differences so far.  I think @Carreau already captured its contents, but it's worth referring to it.

**Comment by minrk** [comment]
@fperez Yup, I didn't mean it was new to IPython, but new to me. I'm always discovering new things! (and `/autocall` is not in our 'beyond plain Python' notebook, but perhaps it should be).

**Comment by fperez** [comment]
Yes, I know :)

I think I didn't put them there b/c I was focusing more on stuff for the notebook, and I think those are more used at the terminal.  In the notebook, I think the cost of calling `foo()` explicitly is worth paying for the increased readability and clarity in the long run, whereas in the terminal, all those shortcuts like `/` and `,` can be handy, if not strictly necessary.

**Comment by Carreau** [comment]
`/...`,`,...` and `;...` are covered here: http://ipython.org/ipython-doc/3/interactive/reference.html#automatic-parentheses-and-quotes : 
```
You can force automatic quoting of a function’s arguments by using , or ; as the first character of a line. For example:

In [1]: ,my_function /home/me  # becomes my_function("/home/me")
If you use ‘;’ the whole argument is quoted as a single string, while ‘,’ splits on whitespace:

In [2]: ,my_function a b c    # becomes my_function("a","b","c")

In [3]: ;my_function a b c    # becomes my_function("a b c")
Note that the ‘,’ or ‘;’ MUST be the first character on the line! This won’t work:

In [4]: x = ,my_function /home/me # syntax error
```

**Pull request merged**